incubator/virtualcluster/experiment/pkg/scheduler/cache/cluster.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+type Cluster struct {
+	name     string
+	labels   map[string]string
+	capacity v1.ResourceList
+
+	alloc  v1.ResourceList
+	slices map[string][]*Slice            // ns key -> slice array
+	pods   map[string]map[string]struct{} // ns key -> Pod map
+}
+
+func NewCluster(name string, labels map[string]string, capacity v1.ResourceList) *Cluster {
+	alloc := capacity.DeepCopy()
+
+	for k, v := range alloc {
+		v.Set(0)
+		alloc[k] = v
+	}
+
+	return &Cluster{
+		name:     name,
+		labels:   labels,
+		capacity: capacity,
+		alloc:    alloc,
+		slices:   make(map[string][]*Slice),
+		pods:     make(map[string]map[string]struct{}),
+	}
+}
+
+func (c *Cluster) DeepCopy() *Cluster {
+	labelcopy := make(map[string]string)
+	for k, v := range c.labels {
+		labelcopy[k] = v
+	}
+
+	out := NewCluster(c.name, c.labels, c.capacity.DeepCopy())
+
+	slicesCopy := make(map[string][]*Slice)
+
+	for k, v := range c.slices {
+		s := make([]*Slice, len(v))
+		for _, each := range v {
+			s = append(s, each.DeepCopy())
+		}
+		slicesCopy[k] = s
+	}
+
+	podsCopy := make(map[string]map[string]struct{})
+	for k, v := range c.pods {
+		for name, _ := range v {
+			podsCopy[k][name] = struct{}{}
+		}
+	}
+	out.slices = slicesCopy
+	out.pods = podsCopy
+	return out
+}
+
+func (c *Cluster) AddNamespace(key string, slices []*Slice) error {
+	if _, ok := c.slices[key]; ok {
+		return fmt.Errorf("Namespace key %s is already in cluster %s, cannot add twice", key, c.name)
```

Comment:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings
Error string should not be capitalized or end with punctuation